src/net/discovery/config.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{
+    helpers::{Direction, Identity},
+    net::{
+        discovery::{Error, PeerDiscovery},
+        MpcHelperClient,
+    },
+    protocol::prss,
+};
+use axum::http::uri::{Authority, Scheme, Uri};
+use rand::thread_rng;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::Formatter;
+use std::path::PathBuf;
+use x25519_dalek::PublicKey;
+
+/// Describes just the origin of a url, i.e.: "http\[s\]://\[authority\]", minus the path and
+/// query parameters
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Origin {
+    scheme: Scheme,
+    authority: Authority,
+}
+
+impl From<Origin> for Uri {
+    fn from(origin: Origin) -> Self {
+        Uri::builder()
+            .scheme(origin.scheme)
+            .authority(origin.authority)
+            .path_and_query("")
+            .build()
+            .unwrap()
+    }
+}
+
+impl Serialize for Origin {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!("{}://{}", self.scheme, self.authority))
+    }
+}
+
+impl<'de> Deserialize<'de> for Origin {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct UrlVisitor;
+        impl<'de> serde::de::Visitor<'de> for UrlVisitor {
+            type Value = Origin;
+
+            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+                formatter.write_str("a valid format origin made up of http[s]://<authority>")
+            }
+
+            fn visit_str<E: serde::de::Error>(self, v: &str) -> Result<Self::Value, E> {
+                let uri = v.parse::<Uri>().map_err(|err| E::custom(err.to_string()))?;
+                let parts = uri.into_parts();
+                let scheme = parts.scheme.ok_or_else(|| E::custom("missing scheme"))?;
+                let authority = parts
+                    .authority
+                    .ok_or_else(|| E::custom("missing authority"))?;
+                Ok(Origin { scheme, authority })
+            }
+        }
+        deserializer.deserialize_str(UrlVisitor)
+    }
+}
+
+/// Configuration values relevant for interacting with other helpers
+#[derive(Serialize, Deserialize)]
+struct Peer {
+    origin: Origin,
+    public_key: PublicKey,
+}
+
+/// All config value necessary to discover other peer helpers of the MPC ring
+#[derive(Serialize, Deserialize)]
+struct Config {
+    peers: [Peer; 3],
+}
+
+impl Config {
+    /// Reads config from `file_location`. Expects file to be json format
+    /// # Errors
+    /// if the file does not exist, or is in an invalid format
+    #[allow(dead_code)] // TODO: will use in upcoming PR
+    pub fn new(file_location: PathBuf) -> Result<Self, Error> {
+        use std::{fs::File, io::BufReader};
+
+        let contents = File::open(file_location)?;
+        let buffered = BufReader::new(contents);
+        Ok(serde_json::from_reader(buffered)?)
```

Comment:
@martinthomson I know that you are very cautious about adding new dependencies, what do you think about using toml-based configs? Json technically works fine, but imo it is bad for configuration because of nesting and braces. Bringing toml would require us to import `config` crate...

Reply:
toml is fine (and better than JSON, yes)

Reply:
now using toml